### PR TITLE
#23168: Migrate acosh as a device op

### LINKED ISSUE #23168: Migrate acosh as a device op
### Is your feature request related to a problem? Please describe

_No response_

### Describe the solution you'd like

_No response_

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by mouliraj-mcw**
### Ticket
Link to Github Issue #23168 

### Problem description
Acosh operation have been implemented as a composite version.

### What's changed

Updated acosh as a llk op.
Negative values are expected to return NaN. Due to a wormhole limitation, the golden function was modified to return inf accordingly.

### Performance Results
In main: 
Screenshot 2025-06-12 at 12 15 17 PM
New version:
Screenshot 2025-06-12 at 12 15 39 PM

So it is 93.96% faster compared to previous approach

### Checklist
- [ ] [All post commit CI](https://github.com/tenstorrent/tt-metal/actions/runs/15735254289
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/runs/15735271221)

### CODE REVIEW

**Review by Aswinmcw — CHANGES_REQUESTED** [review]
check failing pipelines

**Review by Aswinmcw — APPROVED** [review]

**Review by nvelickovicTT — APPROVED** [review]

**Comment by nvelickovicTT** [comment]
Please run both WH and BH post commits as well.

**Review by KalaivaniMCW — APPROVED** [review]
Looks good!

**Review by VirdhatchaniKN — APPROVED** [review]

**Review by dchenTT — APPROVED** [review]

**Review by mcw-anasuya — APPROVED** [review]

File: tests/ttnn/unit_tests/operations/eltwise/test_unary.py [hunk 1]
```diff
@@ -782,3 +782,46 @@ def test_unary_trunc_ttnn_opt(input_shapes, device):
 
     assert_with_ulp(output_tensor, golden_tensor)
     assert_with_pcc(ttnn.to_torch(output_tensor), golden_tensor)
+
+
+@pytest.mark.parametrize(
```
Comment (ayerofieiev-tt) [inline tests/ttnn/unit_tests/operations/eltwise/test_unary.py:781]: This only checks the 4D cases. Please diversify from 1D to 5d
If some case does not work, explicitly mark it disabled

File: tests/ttnn/unit_tests/operations/eltwise/test_unary.py [hunk 2]
```diff
@@ -782,3 +782,46 @@ def test_unary_trunc_ttnn_opt(input_shapes, device):
...
+    "input_shapes",
+    (
+        (torch.Size([1, 1, 32, 32])),
+        (torch.Size([1, 1, 320, 384])),
+        (torch.Size([1, 3, 320, 384])),
+    ),
+)
+@pytest.mark.parametrize(
+    "dtype, low, high",
+    [(ttnn.float32, 1, 100), (ttnn.bfloat16, -100, 1), (ttnn.float32, -100, 1)],
```
Comment (ayerofieiev-tt) [inline tests/ttnn/unit_tests/operations/eltwise/test_unary.py]: Int not supported?

Comment (ayerofieiev-tt) [reply tests/ttnn/unit_tests/operations/eltwise/test_unary.py]: What about bfloat8_b?

File: tests/ttnn/unit_tests/operations/eltwise/test_unary.py [hunk 3]
```diff
@@ -782,3 +782,46 @@ def test_unary_trunc_ttnn_opt(input_shapes, device):
...
+)
+def test_unary_acosh_edge_case_ttnn(input_shapes, device, dtype, low, high):
+    in_data1, input_tensor1 = data_gen_with_range_dtype(input_shapes, low, high, device, ttnn_dtype=dtype)
+
+    output_tensor = ttnn.acosh(input_tensor1)
+    golden_function = ttnn.get_golden_function(ttnn.acosh)
+    golden_tensor = golden_function(in_data1, device=device)
+
+    comp_pass = compare_pcc([output_tensor], [golden_tensor])
```
Comment (ayerofieiev-tt) [inline tests/ttnn/unit_tests/operations/eltwise/test_unary.py]: Why do we still use pcc comparison after the team shared that this accuracy check is not sufficient?

Please check with @patrickroberts

File: tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/llk_math_eltwise_unary_sfpu_trigonometry.h [hunk 4]
```diff
@@ -70,6 +70,18 @@ inline void llk_math_eltwise_unary_sfpu_acos(uint dst_index, int vector_mode = (
     llk_math_eltwise_unary_sfpu_params<APPROXIMATE>(ckernel::sfpu::calculate_acos<APPROXIMATE>, dst_index, vector_mode);
 }
 
+// acosh
+template <bool APPROXIMATE>
+inline void llk_math_eltwise_unary_sfpu_acosh_init() {
+    llk_math_eltwise_unary_sfpu_init<SfpuType::acosh, APPROXIMATE>(ckernel::sfpu::_init_acosh_<APPROXIMATE>());
+}
+
+template <bool APPROXIMATE, int ITERATIONS = 8>
```
Comment (ayerofieiev-tt) [inline tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/llk_math_eltwise_unary_sfpu_trigonometry.h:79]: I know this is used typically, but I am not well informed. 
How do we find the suitable iterations count here?

File: tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/llk_math_eltwise_unary_sfpu_trigonometry.h [hunk 5]
```diff
@@ -70,6 +70,18 @@ inline void llk_math_eltwise_unary_sfpu_acos(uint dst_index, int vector_mode = (
...
+inline void llk_math_eltwise_unary_sfpu_acosh(uint dst_index, int vector_mode = (int)VectorMode::RC) {
```
Comment (ayerofieiev-tt) [inline tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/llk_math_eltwise_unary_sfpu_trigonometry.h:80]: Why should vector_mode be int if it could be of a VectorMode type?

File: ttnn/ttnn/operations/unary.py [hunk 6]
```diff
@@ -200,9 +200,8 @@ def _golden_function_acos(input_tensor_a, *args, device, **kwargs):
 def _golden_function_acosh(input_tensor_a, *args, device, **kwargs):
     import torch
 
-    return torch.nan_to_num(
-        torch.acosh(input_tensor_a), nan=device.sfpu_nan(), posinf=device.sfpu_inf(), neginf=-device.sfpu_inf()
-    )
+    result = torch.acosh(input_tensor_a)
+    return result.masked_fill_(input_tensor_a < 1, float("inf")) if input_tensor_a.dtype == torch.bfloat16 else result
```
Comment (ayerofieiev-tt) [inline ttnn/ttnn/operations/unary.py:205]: Do you need this on both WH and BH?

Comment (mouliraj-mcw) [reply ttnn/ttnn/operations/unary.py:205]: updated.

File: tests/ttnn/unit_tests/operations/eltwise/test_unary.py (hunk 2 above)
Comment (mouliraj-mcw) [reply tests/ttnn/unit_tests/operations/eltwise/test_unary.py]: Int is not supported, Added test cases for bfloat8_b.

File: tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/llk_math_eltwise_unary_sfpu_trigonometry.h (hunk 4 above)
Comment (mouliraj-mcw) [reply tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/llk_math_eltwise_unary_sfpu_trigonometry.h:79]: From what we heard from LLK team, 8 iterations are needed per face

File: tests/ttnn/unit_tests/operations/eltwise/test_unary.py (hunk 3 above)
Comment (mouliraj-mcw) [reply tests/ttnn/unit_tests/operations/eltwise/test_unary.py]: There’s a separate test for ULP accuracy, also currently, ULP doesn’t handle inf and nan cases (support for this is in progress in [this PR](https://github.com/tenstorrent/tt-metal/pull/23533)). So for these specific cases, I’ve excluded ULP testing and used PCC instead. Also for exp and log dependent ops we can't completely replace the pcc test  with ulp tests. when we improve the accuracy of exp and log we will retest the dependant ops with ulp tests.

File: tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/llk_math_eltwise_unary_sfpu_trigonometry.h (hunk 5 above)
Comment (mouliraj-mcw) [reply tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/llk_math_eltwise_unary_sfpu_trigonometry.h:80]: This how the LLK infra has handled the VectorMode for comparison.

File: tests/ttnn/unit_tests/operations/eltwise/test_unary_fp32.py [hunk 7]
```diff
@@ -113,10 +113,10 @@ def run_unary_test(device, h, w, ttnn_function, pcc=0.9999):
 
     input_tensor = ttnn.from_torch(torch_input_tensor, dtype=ttnn.float32, layout=ttnn.TILE_LAYOUT, device=device)
     output_tensor = ttnn_function(input_tensor)
-    output_tensor = ttnn.to_layout(output_tensor, ttnn.ROW_MAJOR_LAYOUT)
-    output_tensor = ttnn.from_device(output_tensor)
-    output_tensor = ttnn.to_torch(output_tensor)
 
+    output_tensor = ttnn.to_torch(output_tensor)
+    print(output_tensor)
```
Comment (nsextonTT) [inline tests/ttnn/unit_tests/operations/eltwise/test_unary_fp32.py]: Do we want to add these prints to the test?

**Review by ayerofieiev-tt — APPROVED** [review]

File: tests/ttnn/unit_tests/operations/eltwise/test_unary_fp32.py (hunk 7 above)
Comment (mouliraj-mcw) [reply tests/ttnn/unit_tests/operations/eltwise/test_unary_fp32.py]: Removed
